File: AtCoder/ACP/BootCamp4b/Hard100_2/python/B_Counting_of_Trees.py
from collections import Counter
MOD = 998244353
N = int(input())
D = list(map(int, input().split()))


def mod_pow(x, n):
    res = 1
    while n > 0:
        if n & 1:
            res = res * x % MOD
        x = x * x % MOD
        n >>= 1
    return res


if D[0] != 0:
    print(0)
    exit()
C = Counter(D)
if C[0] != 1:
    print(0)
    exit()
res = 1
prv = 0
prv_cnt = 1
for k, v in sorted(C.most_common())[1:]:
    if k - prv != 1:
        res = 0
        break
    res = res * mod_pow(prv_cnt, v) % MOD
    prv = k
    prv_cnt = v
print(res)

File: AtCoder/ACP/BootCamp4b/Hard100_2/cpp/B_Counting_of_Trees.cpp
#include <bits/stdc++.h>
#define MOD 998244353
using namespace std;
typedef long long ll;

ll N;
vector<ll> D;

ll mod_pow(ll x, ll n) {
  ll res = 1;
  while (n > 0) {
    if (n & 1) {
      res = res * x % MOD;
    }
    x = x * x % MOD;
    n >>= 1;
  }
  return res;
}

void solve() {
  if (D[0] != 0) {
    puts("0");
    return;
  }
  map<ll, ll> C;
  for (int i = 0; i < N; ++i) {
    ++C[D[i]];
  }
  if (C[0] != 1) {
    puts("0");
    return;
  }
  ll res = 1;
  ll prv = 0;
  ll prv_cnt = 1;
  for (auto p : C) {
    if (p.first == 0) continue;
    if (p.first - prv != 1) {
      res = 0;
      break;
    }
    res = res * mod_pow(prv_cnt, p.second) % MOD;
    prv = p.first;
    prv_cnt = p.second;
  }
  printf("%lld\n", res);
}

int main() {
  cin >> N;
  D = vector<ll>(N);
  for (int i = 0; i < N; ++i) {
    cin >> D[i];
  }
  solve();
}